File: typescript/tests/webhooks.test.ts
import { describe, it, expect, beforeEach, vi } from "vitest";
import { parseWebhook, isConfirmedStateChange } from "../webhooks/handlers.js";
import { processEvent } from "../webhooks/processor.js";
import { markProcessed, getResourceState, resetStore } from "../webhooks/store.js";
import { PagouHttpClient } from "../src/lib/http.js";
import type { PagouConfig } from "../src/lib/config.js";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const here = dirname(fileURLToPath(import.meta.url));
const loadFixture = (name: string) => JSON.parse(readFileSync(join(here, "fixtures", name), "utf8"));

const config: PagouConfig = {
  environment: "sandbox",
  baseUrl: "https://api.sandbox.pagou.ai",
  apiToken: "test",
  timeoutMs: 1000,
  maxRetries: 0,
};

beforeEach(() => resetStore());

describe("parseWebhook", () => {
  it("routes the transaction family via data.event_type", () => {
    const event = parseWebhook({ id: "evt_1", event: "transaction", data: { id: "tx_1", event_type: "transaction.paid" } });
    expect(event).toMatchObject({ id: "evt_1", family: "transaction", eventType: "transaction.paid", resourceId: "tx_1" });
  });

  it("routes the subscription family", () => {
    const event = parseWebhook({ id: "evt_2", event: "subscription", data: { id: "sub_1", event_type: "subscription.renewed" } });
    expect(event).toMatchObject({ family: "subscription", eventType: "subscription.renewed", resourceId: "sub_1" });
  });

  it("routes the transfer family via top-level type and data.object", () => {
    const event = parseWebhook({ id: "evt_3", type: "payout.transferred", data: { object: { id: "tr_1" } } });
    expect(event).toMatchObject({ family: "transfer", eventType: "payout.transferred", resourceId: "tr_1" });
  });

  it("rejects a body with no event id", () => {
    expect(parseWebhook({ event: "transaction", data: {} })).toEqual({ error: "missing_event_id" });
  });

  it("parses each family fixture without error", () => {
    for (const name of ["webhook.transaction.json", "webhook.subscription.json", "webhook.transfer.json"]) {
      expect("error" in parseWebhook(loadFixture(name))).toBe(false);
    }
  });
});

describe("isConfirmedStateChange", () => {
  it("treats terminal money events as confirmed", () => {
    expect(isConfirmedStateChange("transaction.paid")).toBe(true);
    expect(isConfirmedStateChange("payout.transferred")).toBe(true);
  });
  it("treats informational events as non-confirming", () => {
    expect(isConfirmedStateChange("transaction.created")).toBe(false);
    expect(isConfirmedStateChange("subscription.trial_will_end")).toBe(false);
  });
});

describe("dedupe", () => {
  it("returns true once then false for redeliveries", () => {
    expect(markProcessed("evt_x")).toBe(true);
    expect(markProcessed("evt_x")).toBe(false);
  });
});

describe("processEvent", () => {
  it("reconciles against the API and updates state on a confirmed event", async () => {
    const fetchImpl = vi.fn(async () =>
      new Response(JSON.stringify({ success: true, requestId: "r", data: { status: "paid" } }), {
        headers: { "content-type": "application/json" },
      }),
    );
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    await processEvent(
      { id: "evt_1", family: "transaction", eventType: "transaction.paid", resourceId: "tx_1", raw: {} },
      client,
    );
    expect(fetchImpl).toHaveBeenCalledTimes(1);
    expect(getResourceState("tx_1")).toBe("paid");
  });

  it("does not reconcile or change state on a non-confirming event", async () => {
    const fetchImpl = vi.fn();
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    await processEvent(
      { id: "evt_2", family: "transaction", eventType: "transaction.created", resourceId: "tx_2", raw: {} },
      client,
    );
    expect(fetchImpl).not.toHaveBeenCalled();
    expect(getResourceState("tx_2")).toBeUndefined();
  });

  it("skips reconciliation when the confirmed event has no resource id", async () => {
    const fetchImpl = vi.fn();
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    await processEvent(
      { id: "evt_3", family: "transaction", eventType: "transaction.paid", resourceId: "", raw: {} },
      client,
    );
    expect(fetchImpl).not.toHaveBeenCalled();
  });

  it("leaves state unchanged when the resource is not found", async () => {
    const fetchImpl = vi.fn(async () =>
      new Response(JSON.stringify({ message: "not found" }), { status: 404, headers: { "content-type": "application/json" } }),
    );
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    await processEvent(
      { id: "evt_4", family: "transfer", eventType: "payout.transferred", resourceId: "tr_x", raw: {} },
      client,
    );
    expect(getResourceState("tr_x")).toBeUndefined();
  });
});

File: go/tests/webhooks_test.go
package tests

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"testing"

	"github.com/PagouAi/examples/go/webhooks"
)

func loadFixture(t *testing.T, name string) any {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("fixtures", name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	var out any
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("decode fixture %s: %v", name, err)
	}
	return out
}

func TestParseWebhookTransactionFamily(t *testing.T) {
	event, perr := webhooks.ParseWebhook(map[string]any{
		"id": "evt_1", "event": "transaction",
		"data": map[string]any{"id": "tx_1", "event_type": "transaction.paid"},
	})
	if perr != "" {
		t.Fatalf("parse error: %v", perr)
	}
	if event.Family != webhooks.FamilyTransaction || event.EventType != "transaction.paid" || event.ResourceID != "tx_1" {
		t.Errorf("unexpected event: %+v", event)
	}
}

func TestParseWebhookSubscriptionFamily(t *testing.T) {
	event, _ := webhooks.ParseWebhook(map[string]any{
		"id": "evt_2", "event": "subscription",
		"data": map[string]any{"id": "sub_1", "event_type": "subscription.renewed"},
	})
	if event.Family != webhooks.FamilySubscription || event.EventType != "subscription.renewed" || event.ResourceID != "sub_1" {
		t.Errorf("unexpected event: %+v", event)
	}
}

func TestParseWebhookTransferFamily(t *testing.T) {
	event, _ := webhooks.ParseWebhook(map[string]any{
		"id": "evt_3", "type": "payout.transferred",
		"data": map[string]any{"object": map[string]any{"id": "tr_1"}},
	})
	if event.Family != webhooks.FamilyTransfer || event.EventType != "payout.transferred" || event.ResourceID != "tr_1" {
		t.Errorf("unexpected event: %+v", event)
	}
}

func TestParseWebhookRejectsMissingID(t *testing.T) {
	_, perr := webhooks.ParseWebhook(map[string]any{"event": "transaction", "data": map[string]any{}})
	if perr != webhooks.ErrMissingEventID {
		t.Errorf("parse error = %q, want missing_event_id", perr)
	}
}

func TestParseWebhookFixtures(t *testing.T) {
	for _, name := range []string{"webhook.transaction.json", "webhook.subscription.json", "webhook.transfer.json"} {
		if _, perr := webhooks.ParseWebhook(loadFixture(t, name)); perr != "" {
			t.Errorf("%s: parse error %q", name, perr)
		}
	}
}

func TestIsConfirmedStateChange(t *testing.T) {
	confirmed := []string{"transaction.paid", "payout.transferred"}
	for _, e := range confirmed {
		if !webhooks.IsConfirmedStateChange(e) {
			t.Errorf("%s should be confirmed", e)
		}
	}
	informational := []string{"transaction.created", "subscription.trial_will_end"}
	for _, e := range informational {
		if webhooks.IsConfirmedStateChange(e) {
			t.Errorf("%s should not be confirmed", e)
		}
	}
}

func TestDedupe(t *testing.T) {
	store := webhooks.NewStore()
	if !store.MarkProcessed("evt_x") {
		t.Error("first MarkProcessed = false, want true")
	}
	if store.MarkProcessed("evt_x") {
		t.Error("redelivery MarkProcessed = true, want false")
	}
}

func TestProcessEventReconcilesOnConfirmed(t *testing.T) {
	calls := 0
	client := newClient(testConfig(), func(*http.Request) (*http.Response, error) {
		calls++
		return jsonResponse(200, `{"success":true,"requestId":"r","data":{"status":"paid"}}`, nil), nil
	})
	store := webhooks.NewStore()

	err := webhooks.ProcessEvent(client, store, webhooks.Event{
		ID: "evt_1", Family: webhooks.FamilyTransaction, EventType: "transaction.paid", ResourceID: "tx_1",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if calls != 1 {
		t.Errorf("calls = %d, want 1", calls)
	}
	if state, _ := store.ResourceState("tx_1"); state != "paid" {
		t.Errorf("state = %q, want paid", state)
	}
}

func TestProcessEventIgnoresNonConfirming(t *testing.T) {
	calls := 0
	client := newClient(testConfig(), func(*http.Request) (*http.Response, error) {
		calls++
		return jsonResponse(200, `{"success":true,"requestId":"r","data":{}}`, nil), nil
	})
	store := webhooks.NewStore()

	if err := webhooks.ProcessEvent(client, store, webhooks.Event{
		ID: "evt_2", Family: webhooks.FamilyTransaction, EventType: "transaction.created", ResourceID: "tx_2",
	}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if calls != 0 {
		t.Errorf("calls = %d, want 0", calls)
	}
	if _, ok := store.ResourceState("tx_2"); ok {
		t.Error("state for tx_2 should be unset")
	}
}

func TestProcessEventSkipsWhenNoResourceID(t *testing.T) {
	calls := 0
	client := newClient(testConfig(), func(*http.Request) (*http.Response, error) {
		calls++
		return jsonResponse(200, `{"success":true,"requestId":"r","data":{}}`, nil), nil
	})
	store := webhooks.NewStore()

	if err := webhooks.ProcessEvent(client, store, webhooks.Event{
		ID: "evt_3", Family: webhooks.FamilyTransaction, EventType: "transaction.paid", ResourceID: "",
	}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if calls != 0 {
		t.Errorf("calls = %d, want 0", calls)
	}
}

func TestProcessEventLeavesStateUnchangedOnNotFound(t *testing.T) {
	client := newClient(testConfig(), func(*http.Request) (*http.Response, error) {
		return jsonResponse(404, `{"message":"not found"}`, nil), nil
	})
	store := webhooks.NewStore()

	if err := webhooks.ProcessEvent(client, store, webhooks.Event{
		ID: "evt_4", Family: webhooks.FamilyTransfer, EventType: "payout.transferred", ResourceID: "tr_x",
	}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if _, ok := store.ResourceState("tr_x"); ok {
		t.Error("state for tr_x should be unset")
	}
}
